rllib/algorithms/algorithm.py

```diff
@@ -941,6 +943,199 @@ def duration_fn(num_units_done):
+
+        Note that this default implementation does not do anything beyond
+        merging evaluation_config with the normal trainer config.
+
+        Args:
+            duration_fn: An optional callable taking the already run
+                num episodes as only arg and returning the number of
+                episodes left to run. It's used to find out whether
+                evaluation should continue.
+        """
+        # How many episodes/timesteps do we need to run?
+        # In "auto" mode (only for parallel eval + training): Run as long
+        # as training lasts.
+        unit = self.config["evaluation_duration_unit"]
+        eval_cfg = self.config["evaluation_config"]
+        rollout_len = eval_cfg["rollout_fragment_length"]
+        num_envs = eval_cfg["num_envs_per_worker"]
+        auto = self.config["evaluation_duration"] == "auto"
+        duration = (
+            self.config["evaluation_duration"]
+            if not auto
+            else (self.config["evaluation_num_workers"] or 1)
+            * (1 if unit == "episodes" else rollout_len)
+        )
+
+        # Call the `_before_evaluate` hook.
+        self._before_evaluate()
+
+        # Put weights only once into object store and use same object
+        # ref to synch to all workers.
+        self._evaluation_weights_seq_number += 1
+        weights_ref = ray.put(self.workers.local_worker().get_weights())
+        self._sync_filters_if_needed(
+            from_worker=self.workers.local_worker(),
+            workers=self.evaluation_workers,
+            timeout_seconds=eval_cfg.get(
+                "sync_filters_on_rollout_workers_timeout_s"
+            ),
+        )
+
+        if self.config["custom_eval_function"]:
+            raise ValueError(
+                "`custom_eval_function` not supported in combination "
+                "with `evaluation_with_async_requests=True` config setting!"
+            )
+        if (
+            self.evaluation_workers is None
+            and (
+                self.workers.local_worker().input_reader is None
+                or self.config["evaluation_num_workers"] == 0
+            )
+        ):
+            raise ValueError(
+                "Local evaluation OR evaluation without input reader OR evaluation "
+                "with only a local eval worker not supported in combination "
+                "with `evaluation_with_async_requests=True` config setting!"
+            )
+
+        agent_steps_this_iter = 0
+        env_steps_this_iter = 0
+
+        logger.info(f"Evaluating current policy for {duration} {unit}.")
+
+        all_batches = []
+
+        # Default done-function returns True, whenever num episodes
+        # have been completed.
+        if duration_fn is None:
+            def duration_fn(num_units_done):
+                return duration - num_units_done
+
+        def remote_fn(worker, w_ref, w_seq_no):
+            worker.set_weights(weights=w_ref)
```

Comment:
I think we need to update worker.set_weight() to take w_seq_no as a parameter. so that after weight update happens, it can save and remember this seq_no. next time set_weight is called, if seq_nos match, it can simply skip the weight udpate.
we can give it a default so we don't have to update all the call sites.
the reason being that since remote_fn is executed inside a loop, we don't want to actually set_weight every time if worker's weight is already up-to-date.

Reply:
Yeah, makes sense. I'll add that arg and make it None by default, in which case, the update definitely happens. If not None, we check for the current weight sequence number on the worker. If None, the sequence number on the worker should get erased (b/c we wouldn't know to which new seq-no we are actually updating to).